File: api/src/api/v1/Orders/Cosmos.cs
using Azure.Core;
using common;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using System;

namespace api.v1.Orders;

internal static class CosmosModule
{
    public static string OrdersContainerIdentifier { get; } = nameof(OrdersContainerIdentifier);

    public static void ConfigureOrdersContainer(IHostApplicationBuilder builder)
    {
        ConfigureDatabase(builder);

        builder.Services.TryAddKeyedSingleton(OrdersContainerIdentifier,
                                              (provider, _) => GetOrdersContainer(provider));
    }

    private static Container GetOrdersContainer(IServiceProvider provider)
    {
        var database = provider.GetRequiredService<Database>();
        var configuration = provider.GetRequiredService<IConfiguration>();

        var containerName = configuration.GetValueOrThrow("COSMOS_ORDERS_CONTAINER_NAME");

        return database.GetContainer(containerName);
    }

    private static void ConfigureDatabase(IHostApplicationBuilder builder)
    {
        ConfigureCosmosClient(builder);

        builder.Services.TryAddSingleton(GetDatabase);
    }

    private static Database GetDatabase(IServiceProvider provider)
    {
        var client = provider.GetRequiredService<CosmosClient>();
        var configuration = provider.GetRequiredService<IConfiguration>();

        var databaseName = configuration.GetValueOrThrow("COSMOS_DATABASE_NAME");

        return client.GetDatabase(databaseName);
    }

    private static void ConfigureCosmosClient(IHostApplicationBuilder builder)
    {
        var configuration = builder.Configuration;
        var aspireConnectionName = configuration.GetValue("ASPIRE_COSMOS_CONNECTION_NAME")
                                                .IfNone(string.Empty);

        builder.AddAzureCosmosClient(aspireConnectionName,
                                     settings =>
                                     {
                                         configuration.GetValue("COSMOS_ACCOUNT_ENDPOINT")
                                                      .Iter(endpoint =>
                                                      {
                                                          if (Uri.TryCreate(endpoint, UriKind.Absolute, out var uri))
                                                          {
                                                              settings.AccountEndpoint = uri;

                                                              var tokenCredential = builder.Services.BuildServiceProvider().GetRequiredService<TokenCredential>();
                                                              settings.Credential = tokenCredential;
                                                          }
                                                      });

                                         configuration.GetValue("COSMOS_CONNECTION_STRING")
                                                      .Iter(connectionString => settings.ConnectionString = connectionString);
                                     });
    }

    public static PartitionKey GetPartitionKey(Order order) =>
        GetPartitionKey(order.Id);

    public static PartitionKey GetPartitionKey(OrderId orderId) =>
        new(orderId.ToString());
}

File: api/code/api.fsharp/Cosmos.fs
[<RequireQualifiedAccess>]
module Cosmos

open Aspire.Microsoft.Azure.Cosmos
open Azure.Core
open Microsoft.Azure.Cosmos
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.DependencyInjection
open System
open System.Text.Json
open FSharpPlus
open common

let private configureCosmosSettings configuration provider (settings: MicrosoftAzureCosmosSettings) =
    let getConfigurationValue = Configuration.getValue configuration

    let tryParseUri value =
        match Uri.TryCreate(value, UriKind.Absolute) with
        | true, uri -> Some uri
        | _ -> None

    getConfigurationValue "COSMOS_ACCOUNT_ENDPOINT"
    |> bind tryParseUri
    |> iter (fun endpoint ->
        settings.AccountEndpoint <- endpoint
        let tokenCredential = ServiceProvider.getServiceOrThrow<TokenCredential> provider
        settings.Credential <- tokenCredential)

    getConfigurationValue "COSMOS_CONNECTION_STRING"
    |> Option.iter (fun connectionString -> settings.ConnectionString <- connectionString)

let private configureClientOptions (options: CosmosClientOptions) =
    options.EnableContentResponseOnWrite <- false
    options.UseSystemTextJsonSerializerWithOptions <- JsonSerializerOptions.Web

    options.CosmosClientTelemetryOptions <-
        let options = CosmosClientTelemetryOptions()
        options.DisableDistributedTracing <- false
        options

let private configureClient (builder: IHostApplicationBuilder) =
    Azure.configureTokenCredential builder

    let configuration = builder.Configuration

    let connectionName =
        Configuration.getValue configuration "COSMOS_CONNECTION_NAME"
        |> Option.defaultValue String.Empty

    let configureCosmosSettings =
        configureCosmosSettings configuration (builder.Services.BuildServiceProvider())

    AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true);
    
    builder.AddAzureCosmosClient(connectionName, configureCosmosSettings, configureClientOptions)

let private getDatabase (provider: IServiceProvider) =
    let client = provider.GetRequiredService<CosmosClient>()
    let configuration = provider.GetRequiredService<IConfiguration>()

    let databaseName =
        Configuration.getValueOrThrow configuration "COSMOS_DATABASE_NAME"

    client.GetDatabase(databaseName)

let private configureDatabase builder =
    configureClient builder

    ServiceCollection.tryAddSingleton builder.Services getDatabase

let ordersContainerIdentifier = "orders container"

let private getOrdersContainer (provider: IServiceProvider) =
    let database = provider.GetRequiredService<Database>()
    let configuration = provider.GetRequiredService<IConfiguration>()

    let containerName =
        Configuration.getValueOrThrow configuration "COSMOS_ORDERS_CONTAINER_NAME"

    database.GetContainer(containerName)

let configureOrdersContainer builder =
    configureDatabase builder

    ServiceCollection.tryAddKeyedSingleton ordersContainerIdentifier builder.Services getOrdersContainer
